python/ql/src/Security/CWE-020/HostnameRegexpShared.qll

```diff
@@ -0,0 +1,202 @@
+/**
+ * Provides predicates for reasoning about regular expressions
+ * that match URLs and hostname patterns.
+ */
+
+private import HostnameRegexpSpecific
+
+/**
+ * Holds if the given constant is unlikely to occur in the origin part of a URL.
+ */
+predicate isConstantInvalidInsideOrigin(RegExpConstant term) {
+  // Look for any of these cases:
+  // - A character that can't occur in the origin
+  // - Two dashes in a row
```

Comment:
This one surprises me. What about URLs with [Punycode](https://en.wikipedia.org/wiki/Punycode) such as https://xn--wrdle-vua.dk?

Reply:
@erik-krogh , do you know ^ ?

Reply:
I don't know...  
My best guess is that's it's sufficiently uncommon that we don't need to consider it.  
 
@asgerf introduced the JS implementation as part of [this PR](https://github.com/github/codeql/pull/2252).  
Maybe he knows something?

Reply:
I know double dashes are valid in the hostname, but in practice it's an indicator that this regexp is not used to check a hostname string, at least not in a security-relevant context.